### PR TITLE
Octos Maja Brenda Scrabble

### PR DESCRIPTION
**Pull request by mgraonic**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

Feature | Feedback
--- | ---
What are the advantages of using git when collaboratively working on one code base? | By using git both persons can collaborate and work at the same time.
Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? | We made the Score method a Class method because it does not have to do anything with a particular instance so we didn't require an instance method.
Describe an example of a test you wrote for a _nominal case_. | When we tested a valid word and got the score for it.
Describe an example of a test you wrote for an _edge case_. | When we tested a letter of more than 7 letters.
Does your code use any methods from the `Enumerable` mixin? If so, where and why was it helpful? | Yes, in the self.highest_score_from method in the file Scoring we used the method min to get break a tie when two letters get the same score so with min we were able to compare the words and choose the one with less letters.<br><br>Also, we used the delete_if method in the draw_tiles method in the Tilebag file to delete the letter if its quantity was 0. They were very useful to complete and meet the requirements.
What was one method you and your pair used to debug code? | We used several times the pry gem to find out the problem and debug. We also debugged by hardcoding values to better see where things were going wrong.
Is there a specific piece of code you'd like feedback on? | The self.highest_score_from method. Are there places where we could have used fewer extra data structures? Or made the code shorter?
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | About how useful is to talk to each other about our ideas, and that thinking out loud can also be very useful.

### CODE REVIEW

File: lib/scoring.rb [hunk 1]
```diff
@@ -1,9 +1,93 @@
+require 'pry'
+
 module Scrabble
   class Scoring
+
+    @letters = {
+      "A"=>1,
+      "B"=>3,
+      "C"=>3,
```
Comment (droberts-sea) [inline lib/scoring.rb:9]: I like this choice of data structure. Since it's not going to change during one run of the program, this is the sort of thing that would be great as a constant.

File: lib/scoring.rb [hunk 2]
```diff
@@ -1,9 +1,93 @@
...
+
     def self.highest_score_from(array_of_words)
+      if array_of_words == []
+        return nil
+      elsif array_of_words.length == 1
+        return array_of_words[0]
+      else
+        array_of_words.length.times do
+          @words_scores = {}
+          array_of_words.each do |word|
```
Comment (droberts-sea) [inline lib/scoring.rb:65]: You loop over the list of words twice here! I think the outer loop (line 63) could be removed entirely, since you unconditionally return at the end of it (line 86), so it will never run more than once.

File: lib/scoring.rb [hunk 3]
```diff
@@ -1,9 +1,93 @@
...
     end
 
+
     def self.highest_score_from(array_of_words)
+      if array_of_words == []
+        return nil
+      elsif array_of_words.length == 1
+        return array_of_words[0]
+      else
+        array_of_words.length.times do
```
Comment (droberts-sea) [inline lib/scoring.rb:63]: Ideally this check should be unnecessary, and your loop code should do the right thing if there's only one word.

File: lib/scoring.rb [hunk 4]
```diff
@@ -1,9 +1,93 @@
...
+              tied_words << word
+            end
+          end
+
+          winner = ""
+          tied_words.each do |word|
+            if word.length == 7
+              winner = word
+            else
+              winner = tied_words.min{|word1,word2| word1.length <=> word2.length}
```
Comment (droberts-sea) [inline lib/scoring.rb:83]: On line 83 you're computing the min across _all_ the tied words within the loop of all the tied words. This is probably not what you want - it means you're looking at all the words each time you look at one of the words, an O(n^2) operation. We could rewrite this tiebreaker section as follows:

```ruby
winner = tied_words.find{ |word| word.length == 7 }
if winner.nil?
  # min by length, equivalent to your line 83 but a little more concise
  winner = tied_words.min_by{ |word| word.length }
end
```

File: lib/player.rb [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+      @plays = []
+      @player_score = 0
+      @tiles = []
+    end
+
+    def play(word)
+      if @won == true
+        return false
+      else
+        @plays << word
```
Comment (droberts-sea) [inline lib/player.rb:24]: Since the `if` clause returns from the method, you can leave out the `else`:

```ruby
if @won
  return false
end
@plays << word
# ...
```

File: lib/player.rb [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    end
+
+    def total_score
+      return @player_score
+    end
+
+    def won?
+      if @player_score > 100
+        @won = true
+        return true
```
Comment (droberts-sea) [inline lib/player.rb:37]: The idea of storing `@won` is interesting, but since figuring out whether we've won is relatively easy, I don't think it's worthwhile to cache this value. It would be simpler to recompute it every time, and to call `self.won?` in `play` instead of checking `@won`.

File: lib/tilebag.rb [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+        "P" => 2, "Q" => 1, "R" => 6,
+        "S" => 4, "T" => 6, "U" => 4,
+        "V" => 2, "W" => 2, "X" => 1,
+        "Y" => 2, "Z" => 1}
+      end
+      
+      def draw_tiles(num)
+        num.times do
+          @drawn_tiles = []
+          letter = @default_tiles.keys.sample
```
Comment (droberts-sea) [inline lib/tilebag.rb:25]: I don't think this will work the way you want. Since you're initializing `@drawn_tiles` inside the loop, it'll get reset every time, and your collection of drawn tiles won't ever have more than one letter! Moving line 24 outside the loop should fix this issue.

I also don't like that you're using an instance variable for `@drawn_tiles` - since it's only used in this method, I think a local variable would be more appropriate.

File: lib/tilebag.rb [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+require_relative 'player.rb'
+
+module Scrabble
+  class TileBag
+    attr_reader :default_tiles, :tiles_remaining
+
+    def initialize
+      @default_tiles = {"A" => 9,
+        "B" => 2, "C" => 2,
+        "D" => 4, "E" => 12, "F" => 2,
```
Comment (droberts-sea) [inline lib/tilebag.rb:12]: By this logic, you're as likely to draw an `e` as you are a `z`! It also means that you won't ever get more than one of the same letter, which could be bad if you need to draw the last two tiles from the bag and they're both `i`. However, the way you've set up your data structures, it would be difficult to do it any other way.

Instead, you might keep all the letters in the bag in an array. You could seed the array with code like:

```ruby
def initialize
  letters = {
    # the hash you had before
  }
  @tiles = []
  letters.each do |letter, count|
    count.times do
      @tiles << letter
    end
  end
end
```

Then you could pull a random tile with `sample` and get a nice even distribution.

File: lib/tilebag.rb [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+          @default_tiles[letter] -= 1
+          @default_tiles.delete_if {|ltr, qty| qty == 0}
+          @drawn_tiles << letter[0]
+        end
+        return @drawn_tiles
+      end
+
+      def tiles_remaining
+        @tiles_remaining = @default_tiles.values.inject(:+)
+      end
```
Comment (droberts-sea) [inline lib/tilebag.rb:35]: Good use of `inject`

File: specs/player_spec.rb [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+  describe 'play method' do
+    it 'takes a word and add it to an array' do
+      brenda = Scrabble::Player.new("Brenda")
+
+      brenda.play("line")
+      brenda.play("fish")
+      brenda.play("knife")
+      brenda.play("cookie")
+
+      brenda.plays.include? "knife"
```
Comment (droberts-sea) [inline specs/player_spec.rb:38]: You don't make any assertions here, which means as long as none of these methods raise an error, this test has no way to fail! You probably want to replace line 38 with something like

```ruby
brenda.plays.include?("knife").must_equal true

# or, minitest has a must_include expectation:

brenda.plays.must_include "knife"
```

File: specs/player_spec.rb [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+      sarah.play('alcohol')
+
+      sarah.highest_word_score.must_equal 62
+    end
+  end
+
+  describe 'won method' do
+    it 'returns true if score is > 100' do
+      john = Scrabble::Player.new("John")
+      john.play("squeeze")
```
Comment (droberts-sea) [inline specs/player_spec.rb:95]: I would like to see more precision around the "edge" between winning and not winning. What happens if the player has:

- 99 points
- 100 points
- 101 points

File: specs/scoring_spec.rb [hunk 12]
```diff
@@ -42,21 +43,33 @@
...
+    it 'returns the only word in a length 1 array' do
+      words1 = ["cookie"]
+      Scrabble::Scoring.highest_score_from(words1).must_equal "cookie"
     end
 
-    it 'returns the highest word if there are two words' do
+    it 'returns the highest value word if there are two words' do
+      words2 = ["cookie", "cake"]
+      Scrabble::Scoring.highest_score_from(words2).must_equal "cookie"
     end
```
Comment (droberts-sea) [inline specs/scoring_spec.rb:57]: Like we discussed in class, it would probably be wise to check multiple orderings of these words, that is, both `['dog', 'fish']` and `['fish', 'dog']`. You'd be surprised how often this sort of thing turns up a bug.

File: specs/scoring_spec.rb [hunk 13]
```diff
@@ -42,21 +43,33 @@
...
 
     it 'if tied, prefer a word with 7 letters' do
+      words3 = ["qqqqqx", "aeiould"]
+
+      Scrabble::Scoring.highest_score_from(words3).must_equal "aeiould"
```
Comment (droberts-sea) [inline specs/scoring_spec.rb:62]: Good work on this test.

File: specs/tilebag_spec.rb [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+    it 'returns 98 tiles ' do
+      tile_bag = Scrabble::TileBag.new
+      tile_bag.default_tiles.values.sum.must_equal 98
+    end
+
+  describe 'draw_tiles(num)' do
+    it 'returns a collection of random tiles' do
+      tile_bag = Scrabble::TileBag.new
+      tile_bag.draw_tiles(4).must_be_instance_of Array
+    end
```
Comment (droberts-sea) [inline specs/tilebag_spec.rb:32]: You should check more than that it gives back an array here. For example:

- Is the array the correct length?
- Are all the members letters?
- Are the letters that were returned the same as the ones that were removed from the bag (how would you test this?)

Expanding your tests in this way would have caught that bug with draw_tiles.

File: specs/tilebag_spec.rb [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+
+    it 'removes tiles from the tile bag' do
+      tile_bag = Scrabble::TileBag.new
+      tile_bag.draw_tiles(7)
+      tile_bag.default_tiles.values.sum.must_equal 91
+    end
```
Comment (droberts-sea) [inline specs/tilebag_spec.rb:38]: I would also like to see tests for `draw_tiles` that answer the following questions:

- Am I able to draw the last tile from the bag?
- What happens if I try to draw more tiles than are currently in the bag?
- What if I try to draw 0 or -3 tiles?

**Comment by droberts-sea** [comment]
# Scrabble
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | yes
Both Teammates contributed to the codebase | yes
Regular Commits with meaningful commit messages | yes
**Scoring class** |
`score` method | yes
Uses appropriate data structure to store the letter score | yes
Appropriately handles edge cases | yes
`highest_score_from` method | yes
Appropriately handles edge cases | yes
Test implementations for `highest_score_from` | yes
**Player class** |
Uses an array to keep track of words played | yes
Uses existing code to create `highest_scoring_word` and `highest_word_score` | yes
Returns `true` or `false` from `won?` method | yes
Tests edge cases on the `play(word)` | some missing - see inline
**TileBag class** |
Uses a data structure to keep the set of default tiles in the bag | see inline
`draw_tiles` method uses the parameter to draw a set of random tiles | big bug - see inline
`tiles_remaining` method returns a count | yes
Tests `draw_tiles` edge cases | some missing - see inline

Good job overall! The big learning goals I'm looking to see demonstrated on this project are:

- Building complex logic (tie-breaking logic in `highest_score_from`)
- Test coverage
- Object composition (interaction between `Player` and `TileBag`)

I feel that you've done a good job of demonstrating the first and the third of these. Your test coverage could use some work - several of your methods are missing tests entirely, or are missing interesting cases. I've left some inline comments for you to review, please do so, and keep up the hard work!
